core/src/main/java/org/apache/iceberg/encryption/Ciphers.java

```diff
@@ -96,33 +110,45 @@ public AesGcmDecryptor(byte[] keyBytes) {
     }
 
     public byte[] decrypt(byte[] ciphertext, byte[] aad)  {
-      int plainTextLength = ciphertext.length - GCM_TAG_LENGTH - NONCE_LENGTH;
-      Preconditions.checkState(plainTextLength >= 1,
+      return decrypt(ciphertext, 0, ciphertext.length, aad);
+    }
+
+    public byte[] decrypt(byte[] ciphertext, int ciphertextOffset, int ciphertextLength, byte[] aad)  {
+      Preconditions.checkState(ciphertextLength - GCM_TAG_LENGTH - NONCE_LENGTH >= 1,
           "Cannot decrypt cipher text of length " + ciphertext.length +
           " because text must longer than GCM_TAG_LENGTH + NONCE_LENGTH bytes. Text may not be encrypted" +
           " with AES GCM cipher");
 
       // Get the nonce from ciphertext
       byte[] nonce = new byte[NONCE_LENGTH];
-      System.arraycopy(ciphertext, 0, nonce, 0, NONCE_LENGTH);
+      System.arraycopy(ciphertext, ciphertextOffset, nonce, 0, NONCE_LENGTH);
 
-      byte[] plainText = new byte[plainTextLength];
-      int inputLength = ciphertext.length - NONCE_LENGTH;
+      int inputLength = ciphertextLength - NONCE_LENGTH;
       try {
         GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce);
         cipher.init(Cipher.DECRYPT_MODE, aesKey, spec);
         if (null != aad) {
           cipher.updateAAD(aad);
         }
-        cipher.doFinal(ciphertext, NONCE_LENGTH, inputLength, plainText, 0);
+        return cipher.doFinal(ciphertext, ciphertextOffset + NONCE_LENGTH, inputLength);
       }  catch (AEADBadTagException e) {
         throw new RuntimeException("GCM tag check failed. Possible reasons: wrong decryption key; or corrupt/tampered" +
-            "data. AES GCM doesn't differentiate between these two.. ", e);
+            "data. AES GCM doesn't differentiate between these two.", e);
       } catch (GeneralSecurityException e) {
         throw new RuntimeException("Failed to decrypt", e);
       }
+    }
+  }
 
-      return plainText;
+  static byte[] streamBlockAAD(byte[] fileAadPrefix, int currentBlockIndex) {
+    byte[] blockAAD = Ints.toByteArray(currentBlockIndex);
```

Comment:
I don't think this is specific enough. What is happening here? What is the endianness?

Reply:
per the javadoc, https://github.com/google/guava/blob/master/guava/src/com/google/common/primitives/Ints.java#L300 ,
this is an explicitly big endian operation. But we can also spell it out in a comment above this line in our code - or create our own ~ `intToBytesBigEndian` function. -?

Reply:
Why use big endian when other places use little endian?

Reply:
Agreed, we'll use the same endianness.